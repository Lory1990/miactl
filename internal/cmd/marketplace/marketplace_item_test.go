// Copyright Mia srl
// SPDX-License-Identifier: Apache-2.0
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package marketplace

import (
	"testing"

	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

const (
	// Marketplace item
	MarketplaceItemJSON = `{
		"_id": "1234567890abcdefg",
		"name": "RocketScience 101: Hello Universe Example",
		"description": "A simple Hello Universe example based on Rocket-Launcher's Interstellar Template.",
		"type": "example",
		"imageUrl": "/v2/files/download/rocket-launch-image.png",
		"supportedByImageUrl": "/v2/files/download/rocket-science-logo.png",
		"supportedBy": "NASA's Humor Department",
		"documentation": {
			"type": "markdown",
			"url": "https://raw.githubusercontent.com/rocket-launcher/Interstellar-Hello-Universe-Example/master/README.md"
		},
		"categoryId": "rocketScience",
		"resources": {
			"services": {
				"rocket-science-hello-universe-example": {
					"archiveUrl": "https://github.com/rocket-launcher/Interstellar-Hello-Universe-Example/archive/master.tar.gz",
					"containerPorts": [
						{
							"name": "spaceport",
							"from": 80,
							"to": 3000,
							"protocol": "TCP"
						}
					],
					"type": "template",
					"name": "rocket-science-hello-universe-example",
					"pipelines": {
						"space-station-ci": {
							"path": "/projects/space-station%2Fpipelines-templates/repository/files/console-pipeline%2Frocket-template.gitlab-ci.yml/raw"
						}
					}
				}
			}
		}
	}`
	MarketplaceItemYaml = `---
_id: 1234567890abcdefg
name: 'RocketScience 101: Hello Universe Example'
description: A simple Hello Universe example based on Rocket-Launcher's Interstellar
  Template.
type: example
imageUrl: "/v2/files/download/rocket-launch-image.png"
supportedByImageUrl: "/v2/files/download/rocket-science-logo.png"
supportedBy: NASA's Humor Department
documentation:
  type: markdown
  url: https://raw.githubusercontent.com/rocket-launcher/Interstellar-Hello-Universe-Example/master/README.md
categoryId: rocketScience
resources:
  services:
    rocket-science-hello-universe-example:
      archiveUrl: https://github.com/rocket-launcher/Interstellar-Hello-Universe-Example/archive/master.tar.gz
      containerPorts:
      - name: spaceport
        from: 80
        to: 3000
        protocol: TCP
      type: template
      name: rocket-science-hello-universe-example
      pipelines:
        space-station-ci:
          path: "/projects/space-station%2Fpipelines-templates/repository/files/console-pipeline%2Frocket-template.gitlab-ci.yml/raw"
`
)

func TestJSONParsing(t *testing.T) {
	marketplaceItem, err := UnmarshalMarketplaceItem([]byte(MarketplaceItemJSON))
	require.NoError(t, err)
	assert.NotEmpty(t, marketplaceItem)
	snaps.MatchSnapshot(t, marketplaceItem)
}

func TestMarketplaceItemToJSON(t *testing.T) {
	marketplaceItem, err := UnmarshalMarketplaceItem([]byte(MarketplaceItemJSON))
	require.NoError(t, err)
	assert.NotEmpty(t, marketplaceItem)
	json, err := marketplaceItem.MarshalMarketplaceItem()
	require.NoError(t, err)
	assert.NotEmpty(t, json)
	snaps.MatchJSON(t, json)
}

func TestMarketplaceItemToJSONIndent(t *testing.T) {
	marketplaceItem, err := UnmarshalMarketplaceItem([]byte(MarketplaceItemJSON))
	require.NoError(t, err)
	assert.NotEmpty(t, marketplaceItem)
	json, err := marketplaceItem.MarshalMarketplaceItemIndent()
	require.NoError(t, err)
	assert.NotEmpty(t, json)
	snaps.MatchSnapshot(t, string(json))
}

func TestYAMLParsing(t *testing.T) {
	marketplaceItem, err := UnmarshalMarketplaceItemYaml([]byte(MarketplaceItemYaml))
	require.NoError(t, err)
	assert.NotEmpty(t, marketplaceItem)
	snaps.MatchSnapshot(t, marketplaceItem)
}

func TestMarketplaceItemToYAML(t *testing.T) {
	marketplaceItem, err := UnmarshalMarketplaceItem([]byte(MarketplaceItemJSON))
	require.NoError(t, err)
	assert.NotEmpty(t, marketplaceItem)
	yaml, err := marketplaceItem.MarshalMarketplaceItemYaml()
	require.NoError(t, err)
	assert.NotEmpty(t, yaml)
	snaps.MatchSnapshot(t, string(yaml))
}
