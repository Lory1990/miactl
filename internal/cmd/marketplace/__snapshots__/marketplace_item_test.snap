
[TestMarketplaceItemToJSON - 1]
{
 "_id": "1234567890abcdefg",
 "categoryId": "rocketScience",
 "description": "A simple Hello Universe example based on Rocket-Launcher's Interstellar Template.",
 "documentation": {
  "type": "markdown",
  "url": "https://raw.githubusercontent.com/rocket-launcher/Interstellar-Hello-Universe-Example/master/README.md"
 },
 "imageUrl": "/v2/files/download/rocket-launch-image.png",
 "name": "RocketScience 101: Hello Universe Example",
 "resources": {
  "services": {
   "rocket-science-hello-universe-example": {
    "archiveUrl": "https://github.com/rocket-launcher/Interstellar-Hello-Universe-Example/archive/master.tar.gz",
    "containerPorts": [
     {
      "from": 80,
      "name": "spaceport",
      "protocol": "TCP",
      "to": 3000
     }
    ],
    "type": "template"
   }
  }
 },
 "supportedBy": "NASA's Humor Department",
 "supportedByImageUrl": "/v2/files/download/rocket-science-logo.png",
 "type": "example"
}
---

[TestMarketplaceItemToYAML - 1]
_id: 1234567890abcdefg
categoryId: rocketScience
description: A simple Hello Universe example based on Rocket-Launcher's Interstellar
  Template.
documentation:
  type: markdown
  url: https://raw.githubusercontent.com/rocket-launcher/Interstellar-Hello-Universe-Example/master/README.md
imageUrl: /v2/files/download/rocket-launch-image.png
name: 'RocketScience 101: Hello Universe Example'
resources:
  services:
    rocket-science-hello-universe-example:
      archiveUrl: https://github.com/rocket-launcher/Interstellar-Hello-Universe-Example/archive/master.tar.gz
      containerPorts:
      - from: 80
        name: spaceport
        protocol: TCP
        to: 3000
      type: template
supportedBy: NASA's Humor Department
supportedByImageUrl: /v2/files/download/rocket-science-logo.png
type: example

---

[TestYAMLParsing - 1]
marketplace.MarketplaceItem{
    ID:                        &"1234567890abcdefg",
    Category:                  (*marketplace.Category)(nil),
    CategoryID:                &"rocketScience",
    ComingSoon:                (*bool)(nil),
    ComponentsIDS:             nil,
    Description:               &"A simple Hello Universe example based on Rocket-Launcher's Interstellar Template.",
    Documentation:             &marketplace.Documentation{Type:"markdown", URL:"https://raw.githubusercontent.com/rocket-launcher/Interstellar-Hello-Universe-Example/master/README.md"},
    ImageURL:                  &"/v2/files/download/rocket-launch-image.png",
    Name:                      &"RocketScience 101: Hello Universe Example",
    PublishOnMiaDocumentation: (*bool)(nil),
    ReleaseStage:              (*string)(nil),
    RepositoryURL:             (*string)(nil),
    Resources:                 &marketplace.Resources{
        Collections: {},
        Endpoints:   {},
        Services:    {
            "rocket-science-hello-universe-example": {
                AdditionalContainers: nil,
                ArchiveURL:           &"https://github.com/rocket-launcher/Interstellar-Hello-Universe-Example/archive/master.tar.gz",
                ComponentID:          (*string)(nil),
                ContainerPorts:       {
                    {
                        From:     int(80),
                        Name:     "spaceport",
                        Protocol: &"TCP",
                        To:       int(3000),
                    },
                },
                CustomFilesConfig:                    nil,
                DefaultAnnotations:                   nil,
                DefaultArgs:                          nil,
                DefaultConfigMaps:                    nil,
                DefaultDocumentationPath:             (*string)(nil),
                DefaultEnvironmentVariables:          nil,
                DefaultHeaders:                       nil,
                DefaultLabels:                        nil,
                DefaultLogParser:                     (*string)(nil),
                DefaultMonitoring:                    (*marketplace.ServiceDefaultMonitoring)(nil),
                DefaultProbes:                        (*marketplace.ServiceDefaultProbes)(nil),
                DefaultResources:                     (*marketplace.ServiceDefaultResources)(nil),
                DefaultSecrets:                       nil,
                DefaultTerminationGracePeriodSeconds: (*float64)(nil),
                Type:                                 &"template",
            },
        },
        UnsecretedVariables: {},
    },
    SupportedBy:         &"NASA's Humor Department",
    SupportedByImageURL: &"/v2/files/download/rocket-science-logo.png",
    Type:                &"example",
}
---

[TestJSONParsing - 1]
marketplace.MarketplaceItem{
    ID:                        &"1234567890abcdefg",
    Category:                  (*marketplace.Category)(nil),
    CategoryID:                &"rocketScience",
    ComingSoon:                (*bool)(nil),
    ComponentsIDS:             nil,
    Description:               &"A simple Hello Universe example based on Rocket-Launcher's Interstellar Template.",
    Documentation:             &marketplace.Documentation{Type:"markdown", URL:"https://raw.githubusercontent.com/rocket-launcher/Interstellar-Hello-Universe-Example/master/README.md"},
    ImageURL:                  &"/v2/files/download/rocket-launch-image.png",
    Name:                      &"RocketScience 101: Hello Universe Example",
    PublishOnMiaDocumentation: (*bool)(nil),
    ReleaseStage:              (*string)(nil),
    RepositoryURL:             (*string)(nil),
    Resources:                 &marketplace.Resources{
        Collections: {},
        Endpoints:   {},
        Services:    {
            "rocket-science-hello-universe-example": {
                AdditionalContainers: nil,
                ArchiveURL:           &"https://github.com/rocket-launcher/Interstellar-Hello-Universe-Example/archive/master.tar.gz",
                ComponentID:          (*string)(nil),
                ContainerPorts:       {
                    {
                        From:     float64(80),
                        Name:     "spaceport",
                        Protocol: &"TCP",
                        To:       float64(3000),
                    },
                },
                CustomFilesConfig:                    nil,
                DefaultAnnotations:                   nil,
                DefaultArgs:                          nil,
                DefaultConfigMaps:                    nil,
                DefaultDocumentationPath:             (*string)(nil),
                DefaultEnvironmentVariables:          nil,
                DefaultHeaders:                       nil,
                DefaultLabels:                        nil,
                DefaultLogParser:                     (*string)(nil),
                DefaultMonitoring:                    (*marketplace.ServiceDefaultMonitoring)(nil),
                DefaultProbes:                        (*marketplace.ServiceDefaultProbes)(nil),
                DefaultResources:                     (*marketplace.ServiceDefaultResources)(nil),
                DefaultSecrets:                       nil,
                DefaultTerminationGracePeriodSeconds: (*float64)(nil),
                Type:                                 &"template",
            },
        },
        UnsecretedVariables: {},
    },
    SupportedBy:         &"NASA's Humor Department",
    SupportedByImageURL: &"/v2/files/download/rocket-science-logo.png",
    Type:                &"example",
}
---

[TestMarketplaceItemToJSONIndent - 1]
{
 "_id": "1234567890abcdefg",
 "categoryId": "rocketScience",
 "description": "A simple Hello Universe example based on Rocket-Launcher's Interstellar Template.",
 "documentation": {
  "type": "markdown",
  "url": "https://raw.githubusercontent.com/rocket-launcher/Interstellar-Hello-Universe-Example/master/README.md"
 },
 "imageUrl": "/v2/files/download/rocket-launch-image.png",
 "name": "RocketScience 101: Hello Universe Example",
 "resources": {
  "services": {
   "rocket-science-hello-universe-example": {
    "archiveUrl": "https://github.com/rocket-launcher/Interstellar-Hello-Universe-Example/archive/master.tar.gz",
    "containerPorts": [
     {
      "from": 80,
      "name": "spaceport",
      "protocol": "TCP",
      "to": 3000
     }
    ],
    "type": "template"
   }
  }
 },
 "supportedBy": "NASA's Humor Department",
 "supportedByImageUrl": "/v2/files/download/rocket-science-logo.png",
 "type": "example"
}
---
